libraries/AP_HAL_ChibiOS/Storage.cpp

```diff
@@ -44,64 +44,72 @@ extern const AP_HAL::HAL& hal;
 
 void Storage::_storage_open(void)
 {
-    if (_initialised) {
+    if (_initialisedType != StorageBackend::None) {
         return;
     }
 
-#ifdef USE_POSIX
-    // if we have failed filesystem init don't try again
-    if (log_fd == -1) {
-        return;
-    }
-#endif
-
     _dirty_mask.clearall();
 
 #if HAL_WITH_RAMTRON
-    using_fram = fram.init();
-    if (using_fram) {
-        if (!fram.read(0, _buffer, CH_STORAGE_SIZE)) {
+    if (fram.init()) {
+        if (fram.read(0, _buffer, CH_STORAGE_SIZE) == CH_STORAGE_SIZE) {
+            _save_backup();
+            _initialisedType = StorageBackend::FRAM;
+            hal.console->printf("Initialised Storage type=%d\n", _initialisedType);
             return;
         }
-        _save_backup();
-        _initialised = true;
-        return;
     }
+
+    #if defined(HAL_RAMTRON_NO_FALLBACK) && HAL_RAMTRON_NO_FALLBACK
+        AP_HAL::panic("Unable to init RAMTRON storage");
+    #endif
+
     // allow for FMUv3 with no FRAM chip, fall through to flash storage
-#endif
+#endif // HAL_WITH_RAMTRON
 
 #ifdef STORAGE_FLASH_PAGE
-    // load from storage backend
-    _flash_load();
+        // load from storage backend
+        _flash_load();
+        _save_backup();
+        _initialisedType = StorageBackend::Flash;
 #elif defined(USE_POSIX)
-    // allow for fallback to microSD based storage
-    sdcard_retry();
+        // if we have failed filesystem init don't try again
+        if (log_fd == -1) {
+            return;
+        }
 
-    log_fd = AP::FS().open(HAL_STORAGE_FILE, O_RDWR|O_CREAT);
-    if (log_fd == -1) {
-        ::printf("open failed of " HAL_STORAGE_FILE "\n");
-        return;
-    }
-    int ret = AP::FS().read(log_fd, _buffer, CH_STORAGE_SIZE);
-    if (ret < 0) {
-        ::printf("read failed for " HAL_STORAGE_FILE "\n");
-        AP::FS().close(log_fd);
-        log_fd = -1;
-        return;
-    }
-    // pre-fill to full size
-    if (AP::FS().lseek(log_fd, ret, SEEK_SET) != ret ||
-        AP::FS().write(log_fd, &_buffer[ret], CH_STORAGE_SIZE-ret) != CH_STORAGE_SIZE-ret) {
-        ::printf("setup failed for " HAL_STORAGE_FILE "\n");
-        AP::FS().close(log_fd);
-        log_fd = -1;
-        return;
-    }
-    using_filesystem = true;
+        // allow for fallback to microSD based storage
+        if (sdcard_retry()) {
+            log_fd = AP::FS().open(HAL_STORAGE_FILE, O_RDWR|O_CREAT);
+            if (log_fd == -1) {
+                hal.console->printf("open failed of " HAL_STORAGE_FILE "\n");
```

Comment:
::printf() was used as it works in early startup

Reply:
Ack. As above, reverted to use `::printf()`